### PR TITLE
updated dns zone creation/management

### PR DESCRIPTION
**Pull request by justinjpacheco**
- neutron
    - improved 'networks' data structure
      - less redundant
      - added explicit keys for zone creation and fqdn prefix
      - put network subnets into their own array of hashes
  - powerdns
    - added 'also-notify' to notify dns slaves of updates to zone files
    - added configuration support for dns nameservers
    - added cloud fqdn to zone recordset
    - added new zone: catalog
      - contains a list of all the zones currently configured in pdns
      - used by bind-dns slaves to know which zones to sync
    - zone serial time is now current epoch
    - split network cidr into /24 chunks when creating reverse zones
    - added catalog-zone-manage python script
      - creates/updates catalog zone with zones listed in pdns

### CODE REVIEW

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+
+def nzfsum(zone):
+    dns_name = dns.name.from_text(zone)
+    return hashlib.sha1(dns_name.to_wire()).hexdigest()
+
+
+def synchronize_catalog_zone():
+    # use pdnsutil list-all-zones command to get all the zones
+    # convert that list into an array
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Add "and then " before "convert".

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    all_zones = check_output('pdnsutil list-all-zones'.split(' ')).split()
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Use full path to pdnsutil.

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        zones.append({'zone': zone, 'nzfsum': nzfsum(zone)})
+
+    # parse the jinja2 zone template
+    catalog_zone_tmpl = config.get('DEFAULT', 'zone_template')
+    with open(catalog_zone_tmpl) as tmpl:
+        template = Template(tmpl.read())
+
+    # render the zone file
+    serial = int(time.time())
+    catalog_zone_file = config.get('DEFAULT','zone_file')
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Space missing after the comma.

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    with open(catalog_zone_file, 'w') as f:
+        f.write(template.render(zone=catalog_zone,zones=zones,serial=serial))
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Spaces after commas.

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    # load the zone file using pdnsutil
+    load_zone = "pdnsutil load-zone {zone} {zone_file}"
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Use full path to pdnsutil.

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    load_zone = load_zone.format(zone=catalog_zone,zone_file=catalog_zone_file)
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Spaces. 😉

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    parser.add_argument(
+      "--sync",
+      action='store_true',
+      help="synchronize catalog zone"
+    )
+
+    args = parser.parse_args()
+
+    if len(sys.argv) <= 1:
+      parser.print_usage()
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Indentation starting here looks non-standard. Shouldn't the action of the "if" clause be indented four spaces?

File: chef/cookbooks/bcpc/libraries/utils.rb [hunk 8]
```diff
@@ -248,3 +248,47 @@ def generate_ip_address(source_ip:, network_cidr:)
 
   (host_network | host_id).to_s
 end
+
+# take an IPAddress cidr object and break it up into /24 chunks
```
Comment (comay) [inline chef/cookbooks/bcpc/libraries/utils.rb]: So although /16 chunks would be acceptable (for shorter prefixes), for now we're only going to standardize on /24s uniformally?

File: chef/cookbooks/bcpc/libraries/utils.rb [hunk 9]
```diff
@@ -248,3 +248,47 @@ def generate_ip_address(source_ip:, network_cidr:)
...
+      },
+    ]
+  end
+
+  # 24 - the target cidr prefix will give us the amount of network bits we
+  # can use for the /24 networks
+  network_bits = 24 - cidr.prefix.to_i
+
+  # 2 ^ $network_bits gives us the amount of /24's we can have
+  networks = 2**network_bits
```
Comment (comay) [inline chef/cookbooks/bcpc/libraries/utils.rb]: Spaces around the exponent operator?

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import argparse
+import configparser
+from jinja2 import Template
+from subprocess import call,check_output
+
+config = configparser.ConfigParser()
+config.read('/usr/local/etc/catalog-zone/catalog-zone.conf')
+
+
+def nzfsum(zone):
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: Nit but "nzf" is a bit of an awkward name - it doesn't come from pdns, does it? Perhaps just zonefilesum()?

File: chef/cookbooks/bcpc/recipes/powerdns.rb [hunk 11]
```diff
@@ -82,74 +83,106 @@
...
+cookbook_file '/usr/local/sbin/catalog-zone-manage' do
+  source 'powerdns/catalog-zone-manage'
+  mode '0755'
+end
+
+directory '/usr/local/etc/catalog-zone' do
+  action :create
+end
+
+cookbook_file '/usr/local/etc/catalog-zone/zone.j2' do
```
Comment (comay) [inline chef/cookbooks/bcpc/recipes/powerdns.rb]: Is zone.j2 a configuration file? It seems /usr/local/lib/catalog-zone/zone.j2 might be a more suitable location.

File: chef/cookbooks/bcpc/templates/default/powerdns/reverse-zone.erb [hunk 12]
```diff
@@ -1,16 +1,18 @@
...
     <%= @serial %> ; serial
-    3559 ; refresh
+    3600 ; refresh
     600 ; retry
     86400 ; expire
     3600 ; minimum
 )
 
-<%= @reverse_zone %> IN NS ns1.<%= node['bcpc']['cloud']['domain'] %>.
+<% node['bcpc']['powerdns']['nameservers'].each do |host,ip| %>
```
Comment (comay) [inline chef/cookbooks/bcpc/templates/default/powerdns/reverse-zone.erb]: Space after comma.

File: chef/cookbooks/bcpc/templates/default/powerdns/zone.erb [hunk 13]
```diff
@@ -1,21 +1,30 @@
...
     <%= @serial %> ; serial
-    3559 ; refresh
+    3600 ; refresh
     600 ; retry
     86400 ; expire
     3600 ; minimum
 )
 
-@ IN NS ns1.<%= node['bcpc']['cloud']['domain'] %>.
+<% node['bcpc']['powerdns']['nameservers'].each do |host,ip| %>
```
Comment (comay) [inline chef/cookbooks/bcpc/templates/default/powerdns/zone.erb]: Space after comma.

File: chef/cookbooks/bcpc/templates/default/powerdns/zone.erb [hunk 14]
```diff
@@ -1,21 +1,30 @@
...
+@ IN NS <%= host %>
+<% end %>
+
+<% node['bcpc']['powerdns']['nameservers'].each do |host,ip| %>
```
Comment (comay) [inline chef/cookbooks/bcpc/templates/default/powerdns/zone.erb]: Space after comma.

File: chef/cookbooks/bcpc/recipes/powerdns.rb [hunk 15]
```diff
@@ -82,74 +83,106 @@
...
+
+template '/usr/local/etc/catalog-zone/catalog-zone.conf' do
+  source 'powerdns/catalog-zone.conf.erb'
+
+  zone = "catalog.#{node['bcpc']['cloud']['domain']}"
+
+  variables(
+    zone: zone,
+    zone_file: "#{Chef::Config[:file_cache_path]}/#{zone}.zone",
+    zone_template: '/usr/local/etc/catalog-zone/zone.j2'
```
Comment (comay) [inline chef/cookbooks/bcpc/recipes/powerdns.rb]: Same question as above.

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage (hunk 10 above)
Comment (justinjpacheco) [reply chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: nzf is a bind concept and is the official name of the hash type

File: chef/cookbooks/bcpc/libraries/utils.rb (hunk 8 above)
Comment (justinjpacheco) [reply chef/cookbooks/bcpc/libraries/utils.rb]: all class boundaries are supported (/8 and /16) everything else is split into /24

File: chef/cookbooks/bcpc/libraries/utils.rb (hunk 9 above)
Comment (justinjpacheco) [reply chef/cookbooks/bcpc/libraries/utils.rb]: ruby doesn't like that

File: chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        help="synchronize catalog zone"
+    )
+
+    args = parser.parse_args()
+
+    if len(sys.argv) <= 1:
+        parser.print_usage()
+        sys.exit(1)
+
+    if args.sync:
```
Comment (comay) [inline chef/cookbooks/bcpc/files/default/powerdns/catalog-zone-manage]: I think the try/except block should also be indented four spaces, and the inner sections (synchronize_catalog_zone, print, etc) another four spaces.

**Pull request merged**